File: scripts/beauty and sex.py
import numpy as np
import pandas as pd
import scipy.stats as stats
import pymc as pm
import arviz as az
import matplotlib.pyplot as plt
import seaborn as sns



# number of couples in the first four categories and in the fifth ("very attractive")
n_normal, n_attract = 2700, 300

# proportion of girls in the first four categories and in the fifth ("very attractive")
p_normal, p_attract = 0.48, 0.56
girl_group_normal =  int (np.round(n_normal * p_normal))
girl_group_attract = int (np.round(n_attract * p_attract))

p_pooled = ( girl_group_normal + girl_group_attract ) / (n_normal + n_attract)
se = np.sqrt( p_pooled * (1 - p_pooled) * (1/n_normal + 1/n_attract) )
z = (p_normal - p_attract) / se #-2.62

# The test declares significance even using significance as little as 0.01
z_crit = stats.norm.ppf(q=0.01) #-2.32


# Bayesian analysis
# The variation in the human sex ratio occurs in a very narrow range.
# On average the female ratio is  (49.5%).
# For example, in United States report 48.7% girls among whites and 49.2% among blacks. 
# In general, similar differences of half of a percentage point or less have been found in different studies.
# Differences larger than 0.5% are not reported in literature.
# We model the proportion of girls in the two groups as two independent beta-binomial model and eventually we
# check the posterior of the difference.
# We treat the 0.48 - 51 as a 95% interval
# the expected value is (48 + 51)/2 = 49.5
#We get a/(a+b) = 0.495, hence b = 0.495/0.505  * a = 0.98 * a
# Approximately we represent this informations using a =3000, b =3000 * 1.02.
# Our prior knowledge is equivalent to 6000 observations.
alpha = 3000
beta= alpha * 1.02
lower_0025 = stats.beta.ppf(0.025,  a=alpha, b= beta)
upper_0975 =  stats.beta.ppf(0.975, a=alpha, b= beta)

#posterior distribution of the difference of the the theta parameters
with pm.Model() as binomial_diff:
    #same prior for both groups. a priori we have no idea how the theta could
    #vary among the two groups
    theta_normal  =  pm.Beta('theta_normal',  alpha=alpha, beta=beta)
    theta_attract =  pm.Beta('theta_attract', alpha=alpha, beta=beta)
    
    #we need to write the model using the binomial likelihood
    y_normal   = pm.Binomial ('y_normal',  p=theta_normal,  observed=girl_from_normal, n=n_normal)
    y_attract  = pm.Binomial ('y_attract', p=theta_attract, observed=girl_from_attract, n=n_attract)
    
    #Difference between the two mortality rates is a deterministic variable.
    diff_theta  = pm.Deterministic('diff_theta', theta_attract - theta_normal)
    trace  = pm.sample(return_inferencedata=True)

a= az.summary(trace)
 

File: scripts/beauty_and_sex.R
# A study of American adolescents whose attractiveness on a five-point scale was assessed.
# Years later, many of these survey respondents had children.
# The original analysis compared the sex ratio of children of the “very attractive” parents to all other.
# For example, several years ago a researcher analyzed data from a survey of 3000 Americans
# 56% of the children of parents in the highest attractiveness category were girls, 
# compared to 48% of the children of parents in the other categories.
# Out of the 3000, 10% were labelled in the highest attractiveness category.
# Is the difference significant?

#frequentist test
n1 <- 300
n2 <- 2700
p1 <- .56
p2 <- .48
pbar <- (p1 * n1 + p2 * n2) / (n1 + n2)
sd_err <- sqrt( pbar * ( 1 - pbar) * (1/n1 + 1/n2) )
z <- (p1 - p2) / sd_err # 2.6, significant

#Bayesian test

# The variation in the human sex ratio occurs in a very narrow range. 
# For example, a recent count in the United States reported 48.7% girls among whites 
# and 49.2% among blacks.
# In the world, it varies between 46% and 50%.
# Similar differences of half of a percentage point or less have been found when comparing
# based on factors such as birth order, maternal age, or season of birth. 
# Given that attractiveness is itself only subjectively measured, 
#we would find it hard to believe that any difference between more and less attractive 
#parents could be as large as 0.5%.

#tune the beta
a = 48
get_b <- function (a){
  return ( a * 52/48)
}

quantile(rbeta(n=10000, a, get_b(a)), c(0.025,0.975)) # 0.38 - 0.58 

a = 480
quantile(rbeta(n=10000, a, get_b(a)), c(0.025,0.975)) # 0.45 - 0.51
         
a = 800
b = get_b(a)
quantile(rbeta(n=10000, a, b), c(0.025,0.975)) # 0.455 - 0.505
#prior strenght: 800 + 866 = 1666, comparable to the size of the study

#posterior
post_1 <- c(a + p1 * n1, b + (1 - p1) * n1) 
post_2 <- c(a + p2 * n2, b + (1 - p2) * n2) 

CI_group1 <- qbeta ( c(0.025,0.975), post_1[1], post_1[2])
CI_group2 <- qbeta ( c(0.025,0.975), post_2[1], post_2[2])

diff <- rbeta(n=10000, post_1[1], post_1[2]) - 
    rbeta(n=10000, post_2[1], post_2[2]) 
#CI contains the 0 and hence difference  is not significant
CI_diff <- quantile(diff, c(0.025,0.975))
